### PR TITLE
Conversion of Hogwild SGD to estimator

### PR DESCRIPTION
**Pull request by artidoro**
Ongoing work on converting the trainers to estimators (#754). This PR converts Hogwild SGD  (StochasticGradientDescentClassificationTrainer binary classification trainer).

### CODE REVIEW

File: test/Microsoft.ML.Tests/TrainerEstimators/HogwildSGDTests.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    public partial class TrainerEstimators
+    {
+        [Fact]
+        public void TestEstimatorHogwildSGD()
+        {
+            (IEstimator<ITransformer> pipe, IDataView dataView) = GetBinaryClassificationPipeline();
+            pipe.Append(new StochasticGradientDescentClassificationTrainer(Env, "Features", "Label"));
+            TestEstimatorCore(pipe, dataView);
+            Done();
+        }
```
Comment (sfilipi) [inline test/Microsoft.ML.Tests/TrainerEstimators/HogwildSGDTests.cs]: nit, if there is only one test, you can also keep it on the TrainerEstimators file, to have a smaller amount of files.

File: test/Microsoft.ML.StaticPipelineTesting/Training.cs [hunk 2]
```diff
@@ -749,5 +749,43 @@ public void FastTreeRanking()
...
+            var dataSource = new MultiFileSource(dataPath);
+            var ctx = new BinaryClassificationContext(env);
+
+            var reader = TextLoader.CreateReader(env,
+                c => (label: c.LoadBool(0), features: c.LoadFloat(1, 9)));
+
+            IPredictorWithFeatureWeights<float> pred = null;
+
+            var est = reader.MakeNewEstimator()
+                .Append(r => (r.label, preds: ctx.Trainers.StochasticGradientDescentClassificationTrainer(r.label, r.features,
```
Comment (sfilipi) [inline test/Microsoft.ML.StaticPipelineTesting/Training.cs]: >StochasticGradientDescentClassificationTrainer [](start = 59, length = 46)

is this the right one?

File: src/Microsoft.ML.StandardLearners/Standard/LinearClassificationTrainer.cs [hunk 3]
```diff
@@ -1496,21 +1501,21 @@ protected override BinaryPredictionTransformer<TScalarPredictor> MakeTransformer
...
         {
             [Argument(ArgumentType.Multiple, HelpText = "Loss Function", ShortName = "loss", SortOrder = 50)]
             public ISupportClassificationLossFactory LossFunction = new LogLossFactory();
 
-            [Argument(ArgumentType.AtMostOnce, HelpText = "L2 regularizer constant", ShortName = "l2", SortOrder = 50)]
-            [TGUI(Label = "L2 Regularizer Constant", SuggestedSweeps = "1e-7,5e-7,1e-6,5e-6,1e-5")]
+            [Argument(ArgumentType.AtMostOnce, HelpText = "L2 Regularization constant", ShortName = "l2", SortOrder = 50)]
+            [TGUI(Label = "L2 Regularization Constant", SuggestedSweeps = "1e-7,5e-7,1e-6,5e-6,1e-5")]
             [TlcModule.SweepableDiscreteParam("L2Const", new object[] { 1e-7f, 5e-7f, 1e-6f, 5e-6f, 1e-5f })]
-            public float L2Const = (float)1e-6;
```
Comment (sfilipi) [inline src/Microsoft.ML.StandardLearners/Standard/LinearClassificationTrainer.cs]: >L2Const [](start = 25, length = 7)

double -check with @tomfnley on this change. Think we tend to keep the names same as the names of the variables in the respective papers.

File: src/Microsoft.ML.StandardLearners/Standard/LinearClassificationTrainer.cs [hunk 4]
```diff
@@ -1578,8 +1589,49 @@ internal void Check(IHostEnvironment env)
...
+            // Apply the advanced args, if the user supplied any.
+            _args.FeatureColumn = featureColumn;
+            _args.LabelColumn = labelColumn;
+            _args.WeightColumn = weightColumn;
+            _args.MaxIterations = maxIterations;
+            _args.InitLearningRate = initLearningRate;
+            _args.L2Weight = l2Weight;
+            _args.Check(env);
+
+            _loss = _args.LossFunction.CreateComponent(env);
```
Comment (sfilipi) [inline src/Microsoft.ML.StandardLearners/Standard/LinearClassificationTrainer.cs]: >  _loss = _a [](start = 10, length = 12)

i think Pete exposed the loss function for SDCA. Might want to do the same.

File: src/Microsoft.ML.StandardLearners/Standard/LinearClassificationTrainer.cs [hunk 5]
```diff
@@ -1578,8 +1589,49 @@ internal void Check(IHostEnvironment env)
...
+            _args = new Arguments();
+            advancedSettings?.Invoke(_args);
+
+            // Apply the advanced args, if the user supplied any.
+            _args.FeatureColumn = featureColumn;
+            _args.LabelColumn = labelColumn;
+            _args.WeightColumn = weightColumn;
+            _args.MaxIterations = maxIterations;
+            _args.InitLearningRate = initLearningRate;
+            _args.L2Weight = l2Weight;
```
Comment (sfilipi) [inline src/Microsoft.ML.StandardLearners/Standard/LinearClassificationTrainer.cs]: for fast trees, i am printing a warning to the user to let them know that the values of the arguments are being overridden. Since we give them two places to set them, to avoid, feel like we should warn. 

       [https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.FastTree/FastTree.cs#L169](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.FastTree/FastTree.cs#L169)


Also:
[https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.FastTree/FastTree.cs#L104](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.FastTree/FastTree.cs#L104)

File: src/Microsoft.ML.StandardLearners/Standard/LinearClassificationTrainer.cs (hunk 3 above)
Comment (artidoro) [reply src/Microsoft.ML.StandardLearners/Standard/LinearClassificationTrainer.cs]: Ok will do, thank you!

---
In reply to: [222460420](https://github.com/dotnet/machinelearning/pull/1134#discussion_r222460420) [](ancestors = 222460420)

File: src/Microsoft.ML.StandardLearners/Standard/LinearClassificationStatics.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        /// <param name="l2Weight">The L2 regularization constant.</param>
+        /// <param name="advancedSettings">A delegate to apply all the advanced arguments to the algorithm.</param>
+        /// <param name="onFit">A delegate that is called every time the
+        /// <see cref="Estimator{TTupleInShape, TTupleOutShape, TTransformer}.Fit(DataView{TTupleInShape})"/> method is called on the
+        /// <see cref="Estimator{TTupleInShape, TTupleOutShape, TTransformer}"/> instance created out of this. This delegate will receive
+        /// the linear model that was trained.  Note that this action cannot change the result in any way; it is only a way for the caller to
+        /// be informed about what was learnt.</param>
+        /// <returns>The predicted output.</returns>
+        public static (Scalar<float> score, Scalar<float> probability, Scalar<bool> predictedLabel) StochasticGradientDescentClassificationTrainer(this BinaryClassificationContext.BinaryClassificationTrainers ctx,
+            Scalar<bool> label,
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.StandardLearners/Standard/LinearClassificationStatics.cs]: >Scalar<bool> label, [](start = 12, length = 19)

Can you ask Pete or Tom about this one?
I'm just concern what we have Pisgty only on top of bool label. In other universe we used to support R4 where positive label was everything above zero, and negative is everything else and Key value with restriction to key dimensionality is 2.

Bool value is fine, but what if user have Text label, like "Yes" and "No", I don't think we have currently way to transform them to bool, only ToKey, but this method limit us only to bool.

Comment (Ivanidzo4ka) [reply src/Microsoft.ML.StandardLearners/Standard/LinearClassificationStatics.cs]: See CheckBinaryLabel method.

---
In reply to: [222534640](https://github.com/dotnet/machinelearning/pull/1134#discussion_r222534640) [](ancestors = 222534640)

Comment (artidoro) [reply src/Microsoft.ML.StandardLearners/Standard/LinearClassificationStatics.cs]: Ok I will check with them, and let you know. 

---
In reply to: [222534836](https://github.com/dotnet/machinelearning/pull/1134#discussion_r222534836) [](ancestors = 222534836,222534640)

Comment (artidoro) [reply src/Microsoft.ML.StandardLearners/Standard/LinearClassificationStatics.cs]: I checked with Tom and he said that in Pigsty it is important to be very deliberate about the types, and that we will just allow the labels to be Scalar<bool>. This is a constraint of the static api.

---
In reply to: [222775240](https://github.com/dotnet/machinelearning/pull/1134#discussion_r222775240) [](ancestors = 222775240,222534836,222534640)

File: src/Microsoft.ML.StandardLearners/Standard/LinearClassificationTrainer.cs (hunk 3 above)
Comment (artidoro) [reply src/Microsoft.ML.StandardLearners/Standard/LinearClassificationTrainer.cs]: Checked with Tom and he is fine with the name change, as it seems commonly used in the rest of the codebase.

---
In reply to: [222463578](https://github.com/dotnet/machinelearning/pull/1134#discussion_r222463578) [](ancestors = 222463578,222460420)

File: src/Microsoft.ML.StandardLearners/Standard/LinearClassificationStatic.cs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        /// <summary>
+        ///  Predict a target using a linear binary classification model trained with the <see cref="Microsoft.ML.Runtime.Learners.StochasticGradientDescentClassificationTrainer"/> trainer.
+        /// </summary>
+        /// <param name="ctx">The binary classificaiton context trainer object.</param>
+        /// <param name="label">The name of the label column.</param>
+        /// <param name="features">The name of the feature column.</param>
+        /// <param name="weights">The name for the example weight column.</param>
+        /// <param name="maxIterations">The maximum number of iterations; set to 1 to simulate online learning.</param>
+        /// <param name="initLearningRate">The initial Learning Rate used by SGD.</param>
```
Comment (sfilipi) [inline src/Microsoft.ML.StandardLearners/Standard/LinearClassificationStatic.cs]: >Learning R [](start = 55, length = 10)

nit: no need for title case

Comment (artidoro) [reply src/Microsoft.ML.StandardLearners/Standard/LinearClassificationStatic.cs]: Thank you!

---
In reply to: [222897917](https://github.com/dotnet/machinelearning/pull/1134#discussion_r222897917) [](ancestors = 222897917)
